proposals/p1964.md

```diff
@@ -0,0 +1,356 @@
+    part of an escape sequence;
+-   empty character literals (`''`);
+-   ASCII control codes (0...31), including whitespace characters other than
+    word space (tab, line feed, carriage return, form feed, and vertical tab),
+    except when specified with an escape sequence.
+
+### Types
+
+We will have the types `Char8`, `Char16`, and `Char32` representing code units
+in UTF-8, UTF-16, and UTF-32, but we will not support all code units, but only
+those which map directly to the complete value of a code point. However,
+character literals will use their own types distinct from these:
+
+-   We will support value preserving implicit conversions from character
+    literals to code point or code unit types. In particular, a character
+    literal converts to a `Char8` UTF-8 code unit if it is less than or equal to
+    0x7F, and `Char16` UTF-16 code unit if it is less than or equal to 0xFFFF.
+-   Conversions from string or character literals to a non-value-preserving
+    encoding must be explicit.
+-   Conversions from string literals to Unicode strings are implicit, even
+    though the numeric values of the encoding may change.
+
+We can see whether the particular literal is represented in the variable's type
+by only looking at the types.
+
+```
+let allowed: Char8 = 'a';
+```
+
+The above is allowed because the type of `'a'` is the character literal
+consisting of the single Unicode code point 97, which can be converted to
+`Char8` since 97 is less than or equal to 0x7F.
+
+```
+let error1: Char8 = '😃';
+let error2: Char8 = 'AB';
+```
+
+However these should produce errors. The type of `'😃'` is the character literal
+consisting of the single Unicode code point `0x1F603`, which is greater than
+0x7F. The type of `'AB'` is a character literal that is a sequence of two
+Unicode code points, which is not valid.
+
+Literals `'\n'` and `'\u{A}'` represent the same character and so have the same
+type. However, explicitly converting this character literal to another character
+set might result in a character with a different value, but that still
+represents the newline character.
+
+```
+// 0x15 is the new line character in EBCDIC.
+Assert('\n' as EBCDICChar == EBCDICChar.Make(0x15));
+Assert('\u{A}' as EBCDICChar != EBCDICChar.Make(0x0A));
+```
+
+### Operations
+
+Character literals representing a single code point support the following
+operators:
+
+-   Comparison: `<`, `>`, `<=`, `>=` `==`
+-   Plus: `+`. However the behavior of this operator is different from its use
+    with strings. Rather than a concatenation, this will add the value of the
+    two characters:
+    -   If the `+` is used between a character literal representing a single
+        Unicode code point and an integer literal, this should produce a
+        character literal if the result fits in a Unicode code point.
+    -   If the `+` is used between two character literals this will produce an
+        error.
```

Comment:
The top bullet text doesn't seem to agree with the sub-bullet points...

Maybe:
```suggestion
-   Plus: `+`. This doesn't concatenate, but allows numerically adjusting the value:
    -   Only one operand may be a character literal, the other must be an integer literal.
    -   The result is the character literal whose numeric value is the sum of numeric value of the operands. If that sum is not a valid Unicode code point, it is an error.
```

But there may be even better phrasings.

Reply:
Your rewrite is a lot more restrictive. Your rewrite says the only thing that can be added to a character literal is an integer literal, but that is not very useful or common compared to adding a non-literal integer.